cartographer_ros/cartographer_ros/node.cc

```diff
@@ -428,26 +427,35 @@ bool Node::HandleWriteState(
 }
 
 void Node::FinishAllTrajectories() {
-  {
-    carto::common::MutexLocker lock(&mutex_);
-    for (auto& entry : is_active_trajectory_) {
-      const int trajectory_id = entry.first;
-      if (entry.second) {
-        map_builder_bridge_.FinishTrajectory(trajectory_id);
-        entry.second = false;
-      }
+  carto::common::MutexLocker lock(&mutex_);
+  for (const auto& entry : is_active_trajectory_) {
+    const int trajectory_id = entry.first;
+    if (entry.second) {
+      FinishTrajectoryUnderLock(trajectory_id);
     }
   }
-  map_builder_bridge_.RunFinalOptimization();
 }
 
 void Node::FinishTrajectory(const int trajectory_id) {
+  carto::common::MutexLocker lock(&mutex_);
+  FinishTrajectoryUnderLock(trajectory_id);
+}
+
+void Node::FinishTrajectoryUnderLock(const int trajectory_id) {
```

Comment:
Sure, I can revert the last commit. I went by the rule "three or more, use a helper function". :-)